linlearn/_estimator.py

```diff
@@ -128,6 +132,37 @@ def partial_deriv(j, inner_products, state):
 
             return partial_deriv
 
+    def grad_factory(self):
+        X = self.X
+        y = self.y
+        loss = self.loss
+        deriv_loss = loss.deriv_factory()
+        n_samples = self.n_samples
+
+        if self.fit_intercept:
+
+            @jit(**jit_kwargs)
+            def grad(inner_products, state):
+                gradient = np.zeros(X.shape[1] + 1)
+                deriv = 0.0
+                for i in range(n_samples):
+                    deriv = deriv_loss(y[i], inner_products[i])
+                    gradient[0] = deriv
+                    gradient[1:] += deriv * X[i]
+                return gradient / n_samples
+
+            return grad
+        else:
+
+            @jit(**jit_kwargs)
+            def grad(inner_products, state):
+                gradient = np.zeros(X.shape[1])
+                for i in range(n_samples):
+                    gradient += deriv_loss(y[i], inner_products[i]) * X[i]
+                return gradient / n_samples
+
+            return grad
+
 
 ################################################################
```

Comment:
Computing stochastic gradients is specific to SGD-type solvers. Adding such a method and plugging it into all the estimators poses new memory allocation issues and might be slower since it implies more function calls. Why not just manually compute sample gradients in the solvers that need this as we already do ?

Reply:
Yes but SGD, SVRG etc do need grad_sample somehow...

Reply:
It is only a matter of one or two lines using deriv_loss() and accessing a row from the data X